### PR TITLE
Documentation: Added documentation for blocks defined for NAFNet implementation

### LINKED ISSUE #61: docs: Fill in missing Doctrings for models
## Action Items

- [x] Add references to the doctrings for [`SimpleGate`](https://github.com/TrystAI/restorers/blob/main/restorers/model/nafnet/nafblock.py#L11).
- [x] Elaborate docstrings for [`ChannelAttention`](https://github.com/TrystAI/restorers/blob/main/restorers/model/nafnet/nafblock.py#L37) with references and more details.
- [x] Elaborate docstrings for [`SimplifiedChannelAttention`](https://github.com/TrystAI/restorers/blob/main/restorers/model/nafnet/nafblock.py#L71) with references and more details.
- [x] Elaborate docstrings for [`NAFBlock`](https://github.com/TrystAI/restorers/blob/main/restorers/model/nafnet/nafblock.py#L100) with references and more details.
- [x] Add references and parameters to the docstrings for [`PixelShuffle`](https://github.com/TrystAI/restorers/blob/main/restorers/model/nafnet/nafnet.py#L10).
- [x] Add references and parameters to the docstrings for [`UpScale`](https://github.com/TrystAI/restorers/blob/main/restorers/model/nafnet/nafnet.py#L39).
- [x] Add references to the docstrings for [`NAFNet`](https://github.com/TrystAI/restorers/blob/main/restorers/model/nafnet/nafnet.py#L85).

### PR DESCRIPTION
**Pull request by Samarendra109**
Fixes #61 

Changes proposed by this PR :-

Added references to the doctrings for [SimpleGate](https://github.com/TrystAI/restorers/blob/main/restorers/model/nafnet/nafblock.py#L11).
 Elaborated docstrings for [ChannelAttention](https://github.com/TrystAI/restorers/blob/main/restorers/model/nafnet/nafblock.py#L37) with references and more details.
 Elaborated docstrings for [SimplifiedChannelAttention](https://github.com/TrystAI/restorers/blob/main/restorers/model/nafnet/nafblock.py#L71) with references and more details.
 Elaborated docstrings for [NAFBlock](https://github.com/TrystAI/restorers/blob/main/restorers/model/nafnet/nafblock.py#L100) with references and more details.
 Added references and parameters to the docstrings for [PixelShuffle](https://github.com/TrystAI/restorers/blob/main/restorers/model/nafnet/nafnet.py#L10).
 Added parameters to the docstrings for [UpScale](https://github.com/TrystAI/restorers/blob/main/restorers/model/nafnet/nafnet.py#L39).
 Added references to the docstrings for [NAFNet](https://github.com/TrystAI/restorers/blob/main/restorers/model/nafnet/nafnet.py#L85).

Request for Review: @soumik12345 @SauravMaheshkar

### CODE REVIEW

**Comment by Samarendra109** [comment]
@soumik12345 , @SauravMaheshkar @ariG23498 please review.

**Comment by soumik12345** [comment]
@Samarendra109 can you please change the target branch for this PR to saurav/add-auto-docs?

File: restorers/model/nafnet/nafblock.py [hunk 1]
```diff
@@ -12,6 +12,10 @@ class SimpleGate(keras.layers.Layer):
     """
     Simple Gate
     It splits the input of size (b,h,w,c) into tensors of size (b,h,w,c//factor) and returns their Hadamard product
+
+    Reference: NAFNet Paper (Simple Baselines for Image Restoration)
+    https://www.ecva.net/papers/eccv_2022/papers_ECCV/papers/136670017.pdf
```
Comment (ariG23498) [inline restorers/model/nafnet/nafblock.py]: I am sorry but is there a `abs` version of the paper? I don't like it when I click on a link and it starts downloading a paper.

If there is none, we can proceed with this.

File: restorers/model/nafnet/nafblock.py [hunk 2]
```diff
@@ -12,6 +12,10 @@ class SimpleGate(keras.layers.Layer):
...
+
     Parameters:
         factor: the amount by which the channels are scaled down
```
Comment (ariG23498) [inline restorers/model/nafnet/nafblock.py]: Should we not add type hints to the parameters?

File: restorers/model/nafnet/nafblock.py [hunk 3]
```diff
@@ -72,6 +87,17 @@ class SimplifiedChannelAttention(keras.layers.Layer):
...
+
+    The Squeeze and final rescaling step is identical to the ChannelAttention Layer.
+    But following the philosophy of NAFNet paper, the excitation operation with
+        two conv layers with respective activations are replaced with a single conv
+        block. So the inter channel dependency is learnt but any gate or activation
+        is not used.
+        (Check the paper/doc string of NAFBlock for more details)
+
+    Reference: NAFNet Paper (Simple Baselines for Image Restoration)
+    https://www.ecva.net/papers/eccv_2022/papers_ECCV/papers/136670017.pdf
```
Comment (ariG23498) [inline restorers/model/nafnet/nafblock.py]: Same comment regarding the pdf link of paper.

File: restorers/model/nafnet/nafblock.py [hunk 4]
```diff
@@ -101,6 +127,26 @@ class NAFBlock(keras.layers.Layer):
...
+
+    The authors have the idea that any operations of the form,
+    .. math::
+        f(X) \dot \sigma(g(Y))
+    (where f and g are feature maps and \sigma is activation function)
+    can be simplified to the form
+    .. math::
+        X \dot g(Y)
+    Using this idea, all the nonlinear activations are replaced by
+        a series of Hadamard produces
```
Comment (ariG23498) [inline restorers/model/nafnet/nafblock.py]: This is so cool! I love it when we can render latex. Good job 😃

File: restorers/model/nafnet/nafblock.py [hunk 5]
```diff
@@ -101,6 +127,26 @@ class NAFBlock(keras.layers.Layer):
...
+
+    Reference: NAFNet Paper (Simple Baselines for Image Restoration)
+    https://www.ecva.net/papers/eccv_2022/papers_ECCV/papers/136670017.pdf
```
Comment (ariG23498) [inline restorers/model/nafnet/nafblock.py]: Same comment regarding the pdf link of paper.

**Review by ariG23498 — CHANGES_REQUESTED** [review]
I see there are no type hints in the docs. Do we intentionally drop them? I feel documentations should include type hints and also the shape of tensors if there is a need.

File: restorers/model/nafnet/nafblock.py (hunk 1 above)
Comment (soumik12345) [inline restorers/model/nafnet/nafblock.py]:
```suggestion
    Reference:
    
    1. [Simple Baselines for Image Restoration](https://arxiv.org/abs/2204.04676)
```

File: restorers/model/nafnet/nafblock.py (hunk 2 above)
Comment (soumik12345) [inline restorers/model/nafnet/nafblock.py]:
```suggestion
        factor (Optional[int]): the amount by which the channels are scaled down
```

File: restorers/model/nafnet/nafblock.py [hunk 6]
```diff
@@ -38,6 +42,17 @@ class ChannelAttention(keras.layers.Layer):
     """
     Channel Attention layer
 
+    The block is named Squeeze-and-Excitation block (SE Block) in the original paper.
+    First the input is 'squeezed' across the spatial dimension to generate
+        a channel-wise descriptor.
+    Following that the inter channel dependency is learnt by applying
+        two convolution layers.
+    Then finally, the input is rescaled by a channel-wise multiplication with the
+        output of the excitation operation.
```
Comment (soumik12345) [inline restorers/model/nafnet/nafblock.py]:
```suggestion
    1. First the input is 'squeezed' across the spatial dimension to generate a
        channel-wise descriptor.
    2. Following that the inter channel dependency is learnt by applying
        two convolution layers.
    3. Finally, the input is rescaled by a channel-wise multiplication with the
        output of the excitation operation.
```

File: restorers/model/nafnet/nafblock.py [hunk 7]
```diff
@@ -38,6 +42,17 @@ class ChannelAttention(keras.layers.Layer):
...
+
+    Reference: Squeeze-and-Excitation Networks, Hu et al.
+    https://ieeexplore.ieee.org/document/8578843
```
Comment (soumik12345) [inline restorers/model/nafnet/nafblock.py]:
```suggestion
    Reference:
    
    1. [Squeeze-and-Excitation Networks](https://arxiv.org/abs/1709.01507)
```

File: restorers/model/nafnet/nafblock.py (hunk 3 above)
Comment (soumik12345) [inline restorers/model/nafnet/nafblock.py]:
```suggestion
    Reference:
    
    1. [Simple Baselines for Image Restoration](https://arxiv.org/abs/2204.04676)
```

File: restorers/model/nafnet/nafblock.py (hunk 5 above)
Comment (soumik12345) [inline restorers/model/nafnet/nafblock.py]:
```suggestion
    Reference:
    
    1. [Simple Baselines for Image Restoration](https://arxiv.org/abs/2204.04676)
```

File: restorers/model/nafnet/nafblock.py [hunk 8]
```diff
@@ -101,6 +127,26 @@ class NAFBlock(keras.layers.Layer):
...
+        X \dot g(Y)
+    Using this idea, all the nonlinear activations are replaced by
+        a series of Hadamard produces
+
+    Reference: NAFNet Paper (Simple Baselines for Image Restoration)
+    https://www.ecva.net/papers/eccv_2022/papers_ECCV/papers/136670017.pdf
+
     Parameters:
         input_channels: number of channels in the input (as NAFBlock retains the input size in the output)
         factor: factor by which the channels must be increased before being reduced by simple gate.
```
Comment (soumik12345) [inline restorers/model/nafnet/nafblock.py]:
```suggestion
        input_channels (Optional[int]): number of channels in the input (as NAFBlock retains the input size in the output)
        factor (Optional[float]): factor by which the channels must be increased before being reduced by simple gate.
```

File: restorers/model/nafnet/nafnet.py [hunk 9]
```diff
@@ -20,6 +20,10 @@ class PixelShuffle(keras.layers.Layer):
     )
 
     Wrapper Class for tf.nn.depth_to_space
+    Reference: https://www.tensorflow.org/api_docs/python/tf/nn/depth_to_space
+
+    Parameters:
+        upscale_factor: the factor by which the input's spatial dimensions will be scaled.
```
Comment (soumik12345) [inline restorers/model/nafnet/nafnet.py]:
```suggestion
        upscale_factor (int): the factor by which the input's spatial dimensions will be scaled.
```

File: restorers/model/nafnet/nafnet.py [hunk 10]
```diff
@@ -48,6 +52,10 @@ class UpScale(keras.layers.Layer):
         channels//(pixel_shuffle_factor**2)
     )
     While giving input, make sure that (pixel_shuffle_factor**2) divides channels
+
+    Parameters:
+        channels: number of channels in the input.
+        pixel_shuffle_factor: the factor by which the input's spatial dimensions will be scaled.
```
Comment (soumik12345) [inline restorers/model/nafnet/nafnet.py]:
```suggestion
        channels (int): number of channels in the input.
        pixel_shuffle_factor (int): the factor by which the input's spatial dimensions will be scaled.
```

File: restorers/model/nafnet/nafnet.py [hunk 11]
```diff
@@ -95,6 +103,9 @@ class NAFNet(keras.models.Model):
     to add your own implementation for these blocks. Overwrite get_blocks to use your custom block
     in NAFNet. But make sure to follow the restrictions on these methods and blocks.
 
+    Reference: NAFNet Paper (Simple Baselines for Image Restoration)
+    https://www.ecva.net/papers/eccv_2022/papers_ECCV/papers/136670017.pdf
```
Comment (soumik12345) [inline restorers/model/nafnet/nafnet.py]:
```suggestion
    Reference:
    
    1. [Simple Baselines for Image Restoration](https://arxiv.org/abs/2204.04676)
```

File: restorers/model/nafnet/nafnet.py [hunk 12]
```diff
@@ -95,6 +103,9 @@ class NAFNet(keras.models.Model):
...
+
     Parameters:
         filters: denotes the starting filter size.
         middle_block_num: (int) denotes the number of middle blocks.
```
Comment (soumik12345) [inline restorers/model/nafnet/nafnet.py]:
```suggestion
        filters (Optional[int]): denotes the starting filter size.
        middle_block_num: (Optional[int]) denotes the number of middle blocks.
```

**Review by soumik12345 — CHANGES_REQUESTED** [review]
Hi @Samarendra109
I added a few commit suggestions that align with the feedback from @ariG23498 at https://github.com/TrystAI/restorers/pull/66#pullrequestreview-1358368283

**Action Items:**

1. Please remove the `Optional` type specifications, they are not required.
2. Specify the types for all the parameters in the docstring.

**Comment by Samarendra109** [comment]
@soumik12345 , @ariG23498 I have made the changes. Please have a look

File: restorers/model/nafnet/nafblock.py [hunk 13]
```diff
@@ -12,11 +12,17 @@ class SimpleGate(keras.layers.Layer):
...
     Simple Gate
     It splits the input of size (b,h,w,c) into tensors of size (b,h,w,c//factor) and returns their Hadamard product
+
+    Reference:
+
+    1. [Simple Baselines for Image Restoration](https://arxiv.org/abs/2204.04676)
+
     Parameters:
-        factor: the amount by which the channels are scaled down
+        factor (Optional[int]): the amount by which the channels are scaled down
```
Comment (soumik12345) [inline restorers/model/nafnet/nafblock.py:21]:
```suggestion
        factor (int): the amount by which the channels are scaled down
```

File: restorers/model/nafnet/nafnet.py [hunk 14]
```diff
@@ -95,26 +103,34 @@ class NAFNet(keras.models.Model):
...
+    Reference:
+
+    1. [Simple Baselines for Image Restoration](https://arxiv.org/abs/2204.04676)
+
     Parameters:
-        filters: denotes the starting filter size.
-        middle_block_num: (int) denotes the number of middle blocks.
-            Each middle block is a single NAFBlock unit.
-        encoder_block_nums: (tuple) the tuple size denotes the number of encoder blocks.
+        filters (Optional[int]): denotes the starting filter size.
```
Comment (soumik12345) [inline restorers/model/nafnet/nafnet.py:111]:
```suggestion
        filters (int): denotes the starting filter size.
```

File: restorers/model/nafnet/nafnet.py [hunk 15]
```diff
@@ -95,26 +103,34 @@ class NAFNet(keras.models.Model):
...
+            Default filters' size is 16 .
+        middle_block_num (Optional[int]): denotes the number of middle blocks.
```
Comment (soumik12345) [inline restorers/model/nafnet/nafnet.py:113]:
```suggestion
        middle_block_num (int): denotes the number of middle blocks.
```

**Review by soumik12345 — CHANGES_REQUESTED** [review]
Hi @Samarendra109, a few more suggestions and we are good to merge.

**Comment by Samarendra109** [comment]
@soumik12345 @ariG23498 please have a look. The changes were addressed.

**Review by ariG23498 — APPROVED** [review]
Looks good to go! 🚀

**Pull request merged**